DigitalLearningSolutions.Data/Services/DelegateUploadFileService.cs

```diff
@@ -0,0 +1,193 @@
+            IUserDataService userDataService,
+            IRegistrationDataService registrationDataService
+        )
+        {
+            this.userDataService = userDataService;
+            this.registrationDataService = registrationDataService;
+            this.jobGroupsDataService = jobGroupsDataService;
+        }
+
+        public BulkUploadResult ProcessDelegatesFile(IFormFile file, int centreId, DateTime? welcomeEmailDate)
+        {
+            var table = OpenDelegatesTable(file);
+            return ProcessDelegatesTable(table, centreId, welcomeEmailDate);
+        }
+
+        public BulkUploadResult ProcessDelegatesTable(IXLTable table, int centreId, DateTime? welcomeEmailDate)
+        {
+            var jobGroupIds = jobGroupsDataService.GetJobGroupsAlphabetical().Select(item => item.id).ToList();
+            var delegateRows = table.Rows().Skip(1).Select(row => new DelegateTableRow(table, row)).ToList();
+
+            foreach (var delegateRow in delegateRows)
+            {
+                if (IsDelegateRowInvalid(centreId, delegateRow, jobGroupIds, out var approved))
+                {
+                    continue;
+                }
+
+                if (approved.HasValue)
+                {
+                    UpdateDelegate(delegateRow, approved.Value, centreId);
+                }
+                else
+                {
+                    RegisterDelegate(delegateRow, welcomeEmailDate, centreId);
+                }
+            }
+
+            return new BulkUploadResult(delegateRows);
+        }
+
+        public IXLTable OpenDelegatesTable(IFormFile file)
+        {
+            var workbook = new XLWorkbook(file.OpenReadStream());
+            var worksheet = workbook.Worksheet(DelegateDownloadFileService.DelegatesSheetName);
+            var table = worksheet.Tables.Table(0);
+
+            if (!ValidateHeaders(table))
+            {
+                throw new InvalidHeadersException();
+            }
+
+            return table;
+        }
+
+        private bool IsDelegateRowInvalid(int centreId, DelegateTableRow delegateRow, IEnumerable<int> jobGroupIds, out bool? approved)
+        {
+            approved = null;
+            return !delegateRow.IsValid(jobGroupIds) || !TryGetExistingApprovedStatus(delegateRow, centreId, out approved);
+        }
+
+        private void UpdateDelegate(DelegateTableRow delegateRow, bool approved, int centreId)
+        {
+            var record = new DelegateRecord(delegateRow, centreId, approved);
+            var status = userDataService.UpdateDelegateRecord(record);
+            switch (status)
+            {
+                case 0:
+                    delegateRow.RowStatus = RowStatus.Updated;
+                    break;
+                case 1:
+                    delegateRow.RowStatus = RowStatus.Skipped;
+                    break;
+                case 2:
+                    delegateRow.RowStatus = RowStatus.Registered;
+                    break;

```

Comment:
Does the SP also register the delegate if need be? Why do we need the separate method call for registration? 👁️